### PR TITLE
P20-906: Fix restoration of user session after sign-out

### PR DESCRIPTION
**Pull request by artem-vavilov**
## Summary
Because the session on code.org has a lifetime, every HTTP request commits the session to extend its expiration, even if no actual changes have been made.

This can lead to race conditions when two or more concurrent requests attempt to update the session simultaneously. If one request deletes the session while another tries to commit changes, the latter could restore the session immediately after deletion, causing users to remain signed in after signing out. For more details, see [the investigation document](https://docs.google.com/document/d/1_H7SDk1vPnUr5N7I14FPSoYs2bPPBGUz7KUAIE8X_A0).

The only effective solution in this case would be to prevent simultaneous session updates by locking the session commit process. However, this would adversely impact the overall platform performance. Instead, the frequency of session updates was reduced, and the session will be verified to still exist right before committing changes to the store. To achieve this were implemented the following:
1. Prevented asynchronous AJAX requests from committing session changes unless they have made actual modifications,  resulting in only regular HTTP requests resetting the session’s expiration time.
2. After deleting a session, mark it as deleted for 60 seconds so that concurrent requests can load this status from the store to confirm it was not recently deleted before committing any changes.

## Notes
- The issue can be reproduced only in Chromium-based browsers, likely due to their more aggressive handling of concurrent requests, as well as their caching and pre-fetching behavior.
- To reproduce the issue locally, run the app using a multiprocess or multithreaded server like Puma:
```bash
➜  code-dot-org git:(P20-906/session-fix) ✗ cd dashboard && bundle exec puma -C config/puma.rb
```

## Links
- JIRA task: [P20-906](https://codedotorg.atlassian.net/browse/P20-906)
- Related PR: https://github.com/code-dot-org/code-dot-org/pull/62153
- Investigation doc: https://docs.google.com/document/d/1_H7SDk1vPnUr5N7I14FPSoYs2bPPBGUz7KUAIE8X_A0

### CODE REVIEW

File: dashboard/lib/middlewares/redis_session_store.rb
Comment (daynew) [inline dashboard/lib/middlewares/redis_session_store.rb:85]: This is where the "slow" asynchronous requests double-check that the session is NOT deleted, right?

**Review by daynew — APPROVED** [review]
Took me a while but this makes sense to me. The risky thing to me looks like increased traffic to Redis since requests will now double-check the session is not marked for deletion. However, this double-check only happens if changes are being written to the session. I don't know how often that is. We should merge this and monitor our before & after Redis usage metrics.

File: dashboard/lib/middlewares/redis_session_store.rb
Comment (artem-vavilov) [reply dashboard/lib/middlewares/redis_session_store.rb:61]: It is not particularly slow — assuming we do not consider one ten-thousandth of a second to be slow. However, this fix adds approximately 14% more requests to Redis on heavily loaded pages, such as the teacher's homepage. I consider this a reasonable trade-off, given that locking the session could not only significantly increase the number of requests to Redis but also adversely impact the overall platform performance.

Comment (artem-vavilov) [reply dashboard/lib/middlewares/redis_session_store.rb:85]: It is not particularly slow — assuming we do not consider one ten-thousandth of a second to be slow. However, this fix adds approximately 14% more requests to Redis on heavily loaded pages, such as the teacher's homepage. We can consider this a reasonable trade-off, given that locking the session could not only significantly increase the number of requests to Redis but also adversely impact the overall platform performance.

File: dashboard/lib/middlewares/redis_session_store.rb
Comment (Hamms) [inline dashboard/lib/middlewares/redis_session_store.rb:61]: What would you think about extracting both the existing logic and this new logic into their own modules, and then including them in the `RedisSessionStore` class? Something like https://gist.github.com/Hamms/f78ad3ad9e896f2f9a280961db135be5, although I'm sure you can come up with some better names.

File: dashboard/test/ui/features/step_definitions/account_steps.rb
Comment (Hamms) [inline dashboard/test/ui/features/step_definitions/account_steps.rb]: Two nits:

1. Please add "I", for parallelism with our other steps; ie, `Given I sign in "Teacher" 10 times` rather than `Given sign in "Teacher" 10 times`
2. Could you also move this block farther up in the file, so it's adjacent to the other "I sign in/out" steps [here](https://github.com/code-dot-org/code-dot-org/blob/687f48de593e21423134aeac2b6c212f354004ad/dashboard/test/ui/features/step_definitions/account_steps.rb#L19-L38)?

File: dashboard/test/ui/features/platform/signing_in.feature
Comment (Hamms) [inline dashboard/test/ui/features/platform/signing_in.feature]: I'm still not able to get this test to pass locally with `bundle exec ./runner.rb --verbose --html --local -f features/platform/signing_in.feature` when running dashboard with puma, only when running the single-threaded server, and even then it's pretty flaky.

I'm worried it might also fail in other environments (like the DTT) despite passing in drone. Does it pass consistently for you when running locally? Is there some issue with my local environment that might be causing this? 

```
[platform/signing_in]   # Ensures that the session is properly terminated during sign-out and not restored by concurrent request
[platform/signing_in]   # See: https://docs.google.com/document/d/1_H7SDk1vPnUr5N7I14FPSoYs2bPPBGUz7KUAIE8X_A0
[platform/signing_in]   @no_mobile
[platform/signing_in]   Scenario: Teacher successfully signs in and out 10 consecutive times without any issues # features/platform/signing_in.feature:77
[platform/signing_in]     Given I create a teacher named "Teacher"                                              # features/step_definitions/account_steps.rb:216
[platform/signing_in]     And I sign out                                                                        # features/step_definitions/account_steps.rb:276
[platform/signing_in]     Given sign in "Teacher" 10 times                                                      # features/step_definitions/account_steps.rb:336
[platform/signing_in]
[platform/signing_in]       expected: false
[platform/signing_in]            got: true
[platform/signing_in]
[platform/signing_in]       (compared using ==)
[platform/signing_in]        (RSpec::Expectations::ExpectationNotMetError)
[platform/signing_in]       ./features/step_definitions/steps.rb:876:in `/^element "([^"]*)" is (not )?visible$/'
[platform/signing_in]       ./features/step_definitions/account_steps.rb:348:in `/^sign in "([^"]*)" (\d+) times$/'
[platform/signing_in]       ./features/support/hooks.rb:28:in `block in <top (required)>'
[platform/signing_in]       features/platform/signing_in.feature:80:in `Given sign in "Teacher" 10 times'
[platform/signing_in]
[platform/signing_in] Failing Scenarios:
[platform/signing_in] cucumber features/platform/signing_in.feature:77 # Scenario: Teacher successfully signs in and out 10 consecutive times without any issues
[platform/signing_in]
[platform/signing_in] 7 scenarios (1 failed, 6 passed)
[platform/signing_in] 59 steps (1 failed, 58 passed)
[platform/signing_in] 0m43.443s
```

**Review by Hamms — COMMENTED** [review]
Thank you very much for all your hard work and patience on this!

I quite like the overall plan here. It is unfortunate that we're adding a couple extra queries (and connections) to some high-volume operations; but based on our conversations with the folks at AWS this week, I'm optimistic that through a combination of the optimization you merged yesterday and enabling connection pooling we will get to a place where we have enough extra headroom to take on a few extra requests.

The implementation details, unit testing, and comments all look great to me! I have a few nits and one organizational suggestion, and have been having a problem with the new UI test, but overall I think this is real close.

Note that we will need to wait to merge until after connection pooling is enabled, and ideally will want a couple of days to gather metrics for a before-and-after comparison.

File: dashboard/test/ui/features/platform/signing_in.feature
Comment (artem-vavilov) [reply dashboard/test/ui/features/platform/signing_in.feature]: The test isn’t flaky, it's simply very effective at reproducing the issue because the solution only partially fixes it by reducing its frequency.

I’ll create a more realistic scenario where users don’t immediately sign out after signing in.

File: dashboard/lib/middlewares/redis_session_store.rb
Comment (artem-vavilov) [reply dashboard/lib/middlewares/redis_session_store.rb:61]: That's a good idea, Elijah!

**Review by Hamms — COMMENTED** [review]
This looks great to me, thanks so much for all the attention to detail with this one!

I am still working on getting our connection count under control (see https://github.com/code-dot-org/code-dot-org/pull/62394), so I'm holding off on an official approval until we're ready for this to be merged. But I think once we are, this is good to go!

**Review by Hamms — APPROVED** [review]
Once we have finished our post-Hour-of-Code merging of `staging-next` and `staging`, and the staging branch has been reopened, this should be good to merge!

Thanks again for your patience with this one. I'm happy for us to finally benefit from this functionality!

**Pull request merged**
